src/azul/openapi/validation.py

```diff
@@ -0,0 +1,242 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from copy import (
+    copy,
+)
+import json
+import logging
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    TypeVar,
+)
+
+import attr
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    RequirementError,
+)
+from azul.openapi.schema import (
+    python_type_for,
+)
+from azul.plugins.metadata.hca.transform import (
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, kw_only=False)
+class ValidationError(Exception):
+    message: str
+    name: str
+    value: Any
+    schema: Optional[JSON] = None
+    content: Optional[JSON] = None
+
+    def to_dict(self):
+        """
+        :return: Dictionary of atttributes that are not None
+        """
+        return {
+            attribute.name: self.__getattribute__(attribute.name)
+            for attribute in self.__attrs_attrs__
+            if self.__getattribute__(attribute.name) is not None
+        }
+
+
+V = TypeVar('V', bound='SpecValidator')
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class SpecValidator(ABC):
+    """
+    A base class for OpenAPI specification validators. See
+    https://swagger.io/specification/#parameter-object for details about
+    parameter specification. The sub-classes for this class are broken down
+    by the complexity of the parameter that is being defined.
```

Comment:
I do not understand "broken down by the complexity of the parameter that is being defined"

Reply:
Removed to avoid confusion